src/babel.js

```diff
@@ -40,18 +40,28 @@ export default function({ types: t }) {
 
       const tag = path.get('name')
 
+      const isStyleNode = name === 'style' || name === STYLE_COMPONENT
+      const isComponentNode =
+        type === 'JSXMemberExpression' ||
+        (name && name.charAt(0) === name.charAt(0).toUpperCase())
+
+      const hasClassName = path
+        .get('attributes')
+        .some(
+          attribute =>
+            attribute.get('name').node &&
+            attribute.get('name').node.name === 'className'
+        )
+      const hasBinding = Object.values(path.scope.bindings).some(binding =>
+        binding.referencePaths.some(r => r === tag)
+      )
+
       if (
-        name &&
-        name !== 'style' &&
-        name !== STYLE_COMPONENT &&
-        (name.charAt(0) !== name.charAt(0).toUpperCase() ||
-          Object.values(path.scope.bindings).some(binding =>
-            binding.referencePaths.some(r => r === tag)
-          ))
+        state.className &&
+        !isStyleNode &&
+        (!isComponentNode || hasClassName || hasBinding)
```

Comment:
@bernhardc Could we eliminate the requirement of an explicit `className` for Childs Components?

It would not be bad to be able to define Button this way: 
```
const Button = props => <button className={props.ClassName}>{props.children}</button>
```

and styled it without being explicit on `className`, just inyected `jsx-123`

```
const ButtonStyled = props => (
   <>
    <Button>{props.children}</Button>
    <style jsx>{`
       button {
         color: red;
       }
   `}</style>
   </>
)
```

I have tried it on my fork and it is working really well but it would take more consensus to implement it this way